doozerlib/assembly.py

```diff
@@ -0,0 +1,154 @@
+    found = []
+    next_assembly = assembly
+    while next_assembly and isinstance(releases_config, Model):
+        if next_assembly in found:
+            raise ValueError(f'Infinite recursion in {assembly} detected; {next_assembly} detected twice in chain')
+        found.append(next_assembly)
+        target_assembly = releases_config.releases[next_assembly].assembly
+        next_assembly = target_assembly.basis.assembly
+
+
+def assembly_group_config(releases_config: Model, assembly: str, group_config: Model) -> Model:
+    """
+    Returns a group config based on the assembly information
+    and the input group config.
+    :param releases_config: A Model for releases.yaml.
+    :param assembly: The name of the assembly
+    :param group_config: The group config to merge into a new group config (original Model will not be altered)
+    :param _visited: Keeps track of visited assembly definitions to prevent infinite recursion.
+    """
+    if not assembly or not isinstance(releases_config, Model):
+        return group_config
+
+    _check_recursion(releases_config, assembly)
+    target_assembly = releases_config.releases[assembly].assembly
+
+    if target_assembly.basis.assembly:  # Does this assembly inherit from another?
+        # Recursively apply ancestor assemblies
+        group_config = assembly_group_config(releases_config, target_assembly.basis.assembly, group_config)
+
+    target_assembly_group = target_assembly.group
+    if not target_assembly_group:
+        return group_config
+
+    return Model(dict_to_model=merger(target_assembly_group.primitive(), group_config.primitive()))
+
+
+def assembly_metadata_config(releases_config: Model, assembly: str, meta_type: str, distgit_key: str, meta_config: Model) -> Model:
+    """
+    Returns a group config based on the assembly information
+    and the input group config.
+    :param releases_config: A Model for releases.yaml.
+    :param assembly: The name of the assembly
+    :param meta_type: 'rpm' or 'image'
+    :param distgit_key: The name of the underlying component
+    :param meta_config: The meta's config object
+    :return: Returns a computed config for the metadata (e.g. value for meta.config).
+    """
+    if not assembly or not isinstance(releases_config, Model):
+        return meta_config
+
+    _check_recursion(releases_config, assembly)
+    target_assembly = releases_config.releases[assembly].assembly
+
+    if target_assembly.basis.assembly:  # Does this assembly inherit from another?
+        # Recursive apply ancestor assemblies
+        meta_config = assembly_metadata_config(releases_config, target_assembly.basis.assembly, meta_type, distgit_key, meta_config)
+
+    config_dict = meta_config.primitive()
+
+    component_list = target_assembly.members[f'{meta_type}s']
+    for component_entry in component_list:
+        if component_entry.distgit_key == '*' or component_entry.distgit_key == distgit_key:
```

Comment:
When I was testing assembly metadata pins the artifact NVR with `is:`, I didn't set `metadata`, then `metadata.primitive()` raised an error. Not sure if we need to allow None metadata here.

Reply:
Thanks! Fixed.